main/src/actuation/motors.cpp

```diff
@@ -226,6 +230,94 @@ void workspaceZeroZ() {
 	stepperZ.setAcceleration(maxAccelZ);
 }
 
+void autoTouchWorkspaceZ() {
+	enableStepperZ();
+
+	stepperZ.setSpeed(zeroSpeed_0 * ConvLead);
+	while (digitalRead(LIMIT_MACH_Z0) == HIGH) {
+		stepperZ.runSpeed();
+	}
+
+	stepperZ.move(-retract * ConvLead);
+	while (stepperZ.distanceToGo() != 0) {
+		stepperZ.run();
+	}
+
+	stepperZ.setSpeed(zeroSpeed_1 * ConvLead);
+	while (digitalRead(LIMIT_MACH_Z0) == HIGH) {
+		stepperZ.runSpeed();
+	}
+
+	stepperZ.move(-retract * ConvLead);
+	while (stepperZ.distanceToGo() != 0) {
+		stepperZ.run();
+	}
+
+	driverZ.rms_current(maxAutoTouchCurrent_RMS);
+	driverZ.TCOOLTHRS(0xFFFFF);
+	driverZ.TPWMTHRS(0);
+
+	long maxSteps = zRange * ConvLead;
+	zStepCount = 0;
+	zStepTarget = maxSteps;
+	zStepActive = true;
+	digitalWrite(MOT_DIR_Z, HIGH);
+	Timer1.initialize(autoTouchStepInterval);
+	Timer1.attachInterrupt(zStepISR);
+
+	 uint32_t last_time = millis();
+
+	while (zStepActive) {
+		uint32_t ms = millis();
+		if((ms-last_time) > 100) { //run every 0.1s
+			last_time = ms;
+
+			uint16_t sg = driverZ.SG_RESULT();
+			Serial.print("SG_RESULT: ");
+			Serial.println(sg, DEC);
+			if (sg < autoTouchThreshold) {
+				Serial.println("Touched!");
+				zStepActive = false;
+				break;
+			}
+		}
+
+		if ((digitalRead(BUTT_HANDLE_L) != LOW) || (digitalRead(BUTT_HANDLE_R) != LOW)) {
+			zStepActive = false;
+			break;
+		}
+
+		delay(1); // Allow time for ISR and checks
+	}
+
+	Timer1.detachInterrupt();
+	delay(1); // Allow time for the last step to complete
+	Serial.println("Compensate backslash...");
+	stepperZ.move(autoTouchRetraction* ConvLead);
+	while (stepperZ.distanceToGo() != 0) {
+		stepperZ.run();
+	}
+
+	// restore defaults
+	driverZ.rms_current(maxCurrent_RMS);
+	driverZ.TCOOLTHRS(0);
+	driverZ.TPWMTHRS(0x753);
+}
+
+void acceptAutoTouchWorkspaceZ() {
+	stepperZ.setCurrentPosition(0);
+	maxHeight = stepperZ.currentPosition()*1.0f / ConvLead;
```

Comment:
This is unvalid for the auto-touch approach. Should be like this:
```suggestion
	maxHeight = (zStepCount * 1.0f / ConvLead) - autoTouchRetraction;
	stepperZ.setCurrentPosition(0);
```